File: daily/Arrays/763.Partition-labels.go
package daily

func partitionLabels(s string) []int {
	lastIndex := map[rune]int{}
	for i := 0; i < len(s); i++ {
		lastIndex[rune(s[i])] = i
	}

	var res []int
	start, end := 0, 0
	for i := 0; i < len(s); i++ {
		end = max(end, lastIndex[rune(s[i])])
		if end == i {
			res = append(res, end-start+1)
			start = i + 1
		}
	}
	return res
}

File: daily/Arrays/763.Partition-labels.py
from typing import List


class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        # Record last occurrence of each character
        last = {ch: i for i, ch in enumerate(s)}

        res: List[int] = []
        start = end = 0
        for i, ch in enumerate(s):
            end = max(end, last[ch])
            if i == end:
                res.append(end - start + 1)
                start = i + 1
        return res

